### PR TITLE
Feature/runtime interface selection

### PR DESCRIPTION
**Pull request by PeterHeckel-TEK**
This changes adds runtime selection of the interface with the ECC chip. This is done by having the i2c and uart ports as optional items in the transport struct which get populated based on the device path passed in on initialization. After the initalization, match statements are used on each method to determine which protocol to use.

Additionally, optimizations were made in terms of improving some command durations, and altering how the SWI receives packets from the chip.

### CODE REVIEW

File: src/transport.rs [hunk 1]
```diff
@@ -3,80 +3,140 @@ use std::{thread, time::Duration};
...
+            });
+            
+            Ok(Self {
+                i2c_port: None,
+                i2c_address: address,
+                swi_port: Some(swi_port),
+                protocol: TransportProtocol::Swi
+            })    
+        }
+        else {
```
Comment (mlodhiaTek) [inline src/transport.rs]: Consider checking i2c path here as well, Final else print error

File: src/transport.rs [hunk 2]
```diff
@@ -3,80 +3,140 @@ use std::{thread, time::Duration};
...
 
     pub fn send_sleep(&mut self) {
-        let _ = self.send_buf(self.address, &[1]);
+        match self.protocol {
+            TransportProtocol::I2c => {
+                let _ = self.send_i2c_buf(self.i2c_address, &[1]);
+            } 
+            TransportProtocol::Swi => {
+                let mut sleep_msg = BytesMut::new();
+                sleep_msg.put_u8(0xCC);
```
Comment (mlodhiaTek) [inline src/transport.rs]: Can this be constant ? Why 0xCC ? Is it specified in ECC chip data sheet, if so refer it here ?

File: src/transport.rs [hunk 3]
```diff
@@ -3,80 +3,140 @@ use std::{thread, time::Duration};
...
+                let sleep_encoded = self.encode_uart_to_swi(&sleep_msg);
+        
+                let _ = self.swi_port.as_mut().unwrap().write(&sleep_encoded);
+                thread::sleep( Duration::from_micros(300));
```
Comment (mlodhiaTek) [inline src/transport.rs]: Similar to WAKE_DELAY so consider a constant

File: src/transport.rs [hunk 4]
```diff
@@ -3,80 +3,140 @@ use std::{thread, time::Duration};
...
-        let _ = self.send_buf(0, &[0]);
-        Ok(())
+        match self.protocol {
+            TransportProtocol::I2c =>{
+                let _ = self.send_i2c_buf(0, &[0]);
+                thread::sleep(WAKE_DELAY);
+                Ok(())
+            }
+            TransportProtocol::Swi => {
+                if let Err(_err) = self.swi_port.as_mut().unwrap().set_baud_rate(115_200)
```
Comment (mlodhiaTek) [inline src/transport.rs]: Consider adding these baud rates as constant, so if ECC chip changes requiring different baud rates can be changed in one place. Also refer to datasheet or the device if it specified there.

File: src/transport.rs [hunk 5]
```diff
@@ -3,80 +3,140 @@ use std::{thread, time::Duration};
...
+                {
+                    return Err(Error::timeout());
+                }
+                
+                let _ = self.swi_port.as_mut().unwrap().write(&[0]);
+                
+                thread::sleep(WAKE_DELAY);
+                let _ = self.swi_port.as_mut().unwrap().set_baud_rate(230_400);
```
Comment (mlodhiaTek) [inline src/transport.rs]: Same as above

File: src/transport.rs [hunk 6]
```diff
@@ -88,184 +148,59 @@ impl EccTransport {
...
-    fn recv_buf(&mut self, buf: &mut BytesMut,  serial_port: &mut Box<dyn SerialPort>) -> Result {
-        let mut encoded_msg = BytesMut::new();
-        encoded_msg.resize(ATCA_CMD_SIZE_MAX as usize,0);
-        
+    fn recv_swi_buf(&mut self, buf: &mut BytesMut) -> Result {
+        unsafe { buf.set_len(2) };
+        buf[1] = 0xFF;
+
         let mut transmit_flag = BytesMut::new();
         transmit_flag.put_u8(0x88);
```
Comment (mlodhiaTek) [inline src/transport.rs]: TRANSMIT_FLG_MSG = 0x88

File: src/transport.rs [hunk 7]
```diff
@@ -285,25 +220,27 @@ impl EccTransport {
...
-                if *bit == 0x7F || *bit == 0x7E {
-                    *byte ^= 1;
+
+    fn decode_swi_to_uart(&mut self, buf: &mut [u8]) -> Result {
+        for byte_idx in 0..buf.len() {
+            let mut decoded_byte = 0;
+            let mut bit_mask: u8 = 1;
+
+            while bit_mask != 0 {
+                let mut rx_byte = [0; 1];
```
Comment (mlodhiaTek) [inline src/transport.rs:235]: Should it be [0,1] for array declaration ?

File: src/transport.rs [hunk 8]
```diff
@@ -88,184 +148,59 @@ impl EccTransport {
...
-    }
-
-    fn send_buf(&mut self, buf: &[u8], serial_port: &mut Box<dyn SerialPort>) -> Result {
+    fn send_swi_buf(&mut self, buf: &[u8]) -> Result {
         
-        let send_size = serial_port.write(buf)?;
+        let send_size = self.swi_port.as_mut().unwrap().write(buf)?;
 
         //Each byte takes ~45us to transmit, so we must wait for the transmission to finish before proceeding
         let uart_tx_time = Duration::from_micros( (buf.len() * 45) as u64);
```
Comment (mlodhiaTek) [inline src/transport.rs]: BYTE_DELAY_US = 45

File: src/transport.rs (hunk 7 above)
Comment (PeterHeckel-TEK) [reply src/transport.rs:235]: No this is how rust does array initialization. It means an array of length 1 with all items initialized to 0

Comment (mlodhiaTek) [reply src/transport.rs:235]: Yes you are correct. Just checked it. Thanks for clarification.
https://doc.rust-lang.org/rust-by-example/primitives/array.html

**Review by mlodhiaTek — APPROVED** [review]

File: src/transport.rs [hunk 9]
```diff
@@ -3,80 +3,140 @@ use std::{thread, time::Duration};
...
-impl EccTransport {
+impl EccTransport
+{
     pub fn from_path(path: &str, address: u16) -> Result<Self> {
-        let mut i2c = I2c::from_path(path)?;
-        i2c.smbus_set_slave_address(address, false)?;
-        Ok(Self { i2c, address })
+        
+        if path.contains("/dev/tty"){
+            let swi_port = serialport::new(path, 230_400)
```
Comment (kschluff-tektelic) [inline src/transport.rs]: `path.starts_with("/dev/tty")` might be better.

**Review by kschluff-tektelic — APPROVED** [review]

**Pull request merged**